### PR TITLE
[Request] Normalize malformed SERVER_NAME fallback hosts

### LINKED ISSUE #331: [Request] Host fallback paths do not normalize malformed SERVER_NAME values with an embedded port
## Description:

When `HTTP_HOST` is missing and `SERVER_NAME` already contains `:port`, Rage does not normalize the fallback host value before using it in request-host code paths.

This is an edge case Rack env, but it currently breaks multiple host-based paths on latest `main`.

## MRE:

```ruby
require "domain_name"
require "./lib/rage/all"

request_env = {
  "REQUEST_METHOD" => "GET",
  "rack.url_scheme" => "http",
  "PATH_INFO" => "/",
  "QUERY_STRING" => "",
  "rack.input" => StringIO.new,
  "SERVER_NAME" => "google.com:3000",
  "SERVER_PORT" => "3000"
}

request = Rage::Request.new(request_env)
puts "request.host => #{request.host.inspect}"
puts "request.url => #{request.url.inspect}"
puts "request.domain => #{request.domain.inspect}"

router = Rage::Router::Backend.new
router.on("GET", "/photos", ->(_) { "get photos" }, constraints: { host: "google.com" })
route_env = request_env.merge("PATH_INFO" => "/photos")
handler = router.lookup(route_env)
puts "router.lookup => #{handler.inspect}"

cookie_env = request_env.merge("SERVER_NAME" => "cookie.test.com:3000")
headers = {}

begin
  cookies = Rage::Cookies.new(cookie_env, headers)
  cookies[:x] = { value: 1, domain: :all }
  puts "cookies header => #{headers.inspect}"
rescue => e
  puts "cookies error => #{e.class}: #{e.message}"
end
```

Image

## Current behavior:

- `request.host` is `nil`.
- `request.url` becomes `"http://google.com:3000:3000/"`.
- `request.domain` is `nil`.
- exact host constraints do not match because `router.lookup` returns `nil`.
- cookie domain resolution can raise `TypeError: NilClass is not a String` when `domain: :all` is used.

### Actual output:

```ruby
request.host => nil
request.url => "http://google.com:3000:3000/"
request.domain => nil
router.lookup => nil
cookies error => TypeError: NilClass is not a String
```

## Expected behavior:

- Rage should normalize the fallback host the same way it normalizes `HTTP_HOST`, even if `SERVER_NAME` is malformed and already includes a port.
- `request.host` should be `"google.com"`.
- `request.url` should be `"http://google.com:3000/"`.
- `request.domain` should be `"google.com"`.
- exact host constraints for `"google.com"` should still match.
- cookie domain resolution should derive from `"cookie.test.com"` instead of raising.

### PR DESCRIPTION
**Pull request by RedZapdos123**
## Description:

Issue #331 reported that host fallback paths could break when `HTTP_HOST` was missing and `SERVER_NAME` already included a port, because Rage used the malformed fallback value as-is.

This PR normalizes malformed `SERVER_NAME` fallback hosts before they are used in request host handling, exact host constraints, and cookie domain resolution.

It also adds regression tests for request URL/host/domain handling, router exact host constraints, and cookie domain resolution when `SERVER_NAME` contains an embedded port, and updates `CHANGELOG.md`.

Closes #331.

## Checklist:

- [x] I have run linting checks using `bundle exec rubocop --except Layout/EndOfLine lib/rage/internal.rb lib/rage/request.rb lib/rage/cookies.rb lib/rage/router/constrainer.rb spec/rage/request_spec.rb spec/router/constraints_spec.rb spec/controller/api/cookies_spec.rb`.
- [x] I have run focused tests using `bundle exec rspec spec/rage/request_spec.rb spec/router/constraints_spec.rb spec/controller/api/cookies_spec.rb`.
- [x] I have run broader regression checks in WSL using `bundle exec rspec spec/controller/api spec/rage/request_spec.rb spec/router/constraints_spec.rb`.

### Before the fix:

image

### After the fix:

image

### CODE REVIEW

**Review by rsamoilov — COMMENTED** [review]
I'm a bit hesitant about these changes.

They seem correct overall, but in practice, Iodine rejects requests without the `Host` header before they even reach Rage. So the only way this code is likely to be executed is in some edge-case tests.

Would you suggest to still merge it?

**Comment by RedZapdos123** [comment]
I think it is reasonable to merge because the code changes are narrow and only affects the existing `SERVER_NAME` fallback branch when that fallback value is malformed and already includes a port.

So....This does not change normal requests with `HTTP_HOST`. It also does not change well formed `SERVER_NAME` fallback behavior. It only makes the fallback handling internally consistent for direct Rack env entry points in Rage.

I checked a few cases against the current main and this branch:

- direct `Rage::Request.new(env)`:
  - before: `request.host => nil`, `request.url => "http://google.com:3000:3000/"`, `request.domain => nil`.
  - after: host/url/domain normalize correctly.

- Direct exact host lookup through `Rage::Router::Backend#lookup(env)`:
  - before: the host-constrained lookup returned `nil`.
  - after: it matches correctly.

- Direct `Rage::Cookies.new(env, headers)`:
  - Before: `domain: :all` raised `TypeError: NilClass is not a String`.
  - After: it resolves `domain=test.com` correctly.

- `Rage::Cable::Connection#request`:
  - Before: `cable.request.host => nil` and `cable.request.url` had the duplicated port.
  - After: both normalize correctly.

Though the real frequency of such cases maybe low, but given that the current behavior can still produce `nil` hosts, duplicated ports in `request.url`, failed exact host matches, and cookie-domain errors in that fallback branch. I think this PR is a good, low risk fix for these already existing direct Rack env code paths.

File: lib/rage/cookies.rb [hunk 1]
```diff
@@ -202,6 +202,7 @@ def []=(key, value)
 
     if (domain = value[:domain])
       host = Rack::Request.new(@env).host
+      host ||= Rage::Internal.extract_host(@env["SERVER_NAME"]) unless @env["HTTP_HOST"]
```
Comment (rsamoilov) [inline lib/rage/cookies.rb]: Can we instead use `Rage::Request` here?

File: lib/rage/internal.rb [hunk 2]
```diff
@@ -44,6 +44,19 @@ def build_arguments(method, arguments)
       }.join(", ")
     end
 
+    # Extract the host from a host:port authority while leaving bare IPv6 literals unchanged.
+    # @param authority [String, nil]
+    # @return [String, nil]
+    def extract_host(authority)
```
Comment (rsamoilov) [inline lib/rage/internal.rb]: I think `Rage::Request` would be a better place for this method.

Comment (RedZapdos123) [reply lib/rage/internal.rb]: I moved the host normalization helper into `Rage::Request` so it stays with the request abstraction instead of `Rage::Internal`.

File: lib/rage/cookies.rb (hunk 1 above)
Comment (RedZapdos123) [reply lib/rage/cookies.rb]: Yes. Thank You for the suggestion.

I changed this to use `Rage::Request.new(@env).host` so cookies reuse the same normalized request host path instead of duplicating the fallback handling here.

**Comment by rsamoilov** [comment]
Thank you for opening the PR. Unfortunately, I'm inclined to close the PR for several reasons:

* Iodine rejects requests without the Host header before they even reach Rage
* `google.com:3000` is not a valid `SERVER_NAME` according to the Rack spec
* I couldn't reproduce a real-world workflow where the fallback would work
